src/herder/test/HerderTests.cpp

```diff
@@ -1519,12 +1526,146 @@ TEST_CASE("surge pricing", "[herder][txset]")
+            TxSetFrameConstPtr txSet = TxSetFrame::makeFromTransactions(
+                TxSetFrame::TxPhases{{tx}, {sorobanTx}}, *app, 0, 0,
+                invalidPhases);
+
+            // Everything fits
+            for (auto const& phase : invalidPhases)
+            {
+                REQUIRE(phase.empty());
+            }
+            REQUIRE(txSet->sizeTx() == 2);
+        }
+        SECTION("soroban surge pricing, classic unaffected")
+        {
+            // Another soroban tx with higher fee, which will be selected
+            auto sorobanTxHighFee = createSimpleDeployContractTx(
+                *app, acc3, baseFee * 2, /* refundableFee */ 1200, resources);
+            TxSetFrame::TxPhases invalidPhases;
+            invalidPhases.resize(2);
+            TxSetFrameConstPtr txSet = TxSetFrame::makeFromTransactions(
+                TxSetFrame::TxPhases{{tx}, {sorobanTx, sorobanTxHighFee}}, *app,
+                0, 0, invalidPhases);
+
+            for (auto const& phase : invalidPhases)
+            {
+                REQUIRE(phase.empty());
+            }
+            REQUIRE(txSet->sizeTx() == 2);
+            for (auto const& tx : txSet->getTxs())
+            {
+                // sorobanTx got kicked out in favor of sorobanTxHighFee
+                REQUIRE(tx != sorobanTx);
+            }
+        }
+        SECTION("soroban surge pricing with gap")
+        {
+            // Another soroban tx with high fee and a bit less resources
+            // Still half capacity available
+            resources.readBytes = conf.txMaxReadBytes() / 2;
+            auto sorobanTxHighFee = createSimpleDeployContractTx(
+                *app, acc3, baseFee * 2, /* refundableFee */ 1200, resources);
+
+            // Create another small soroban tx, with small fee. It should be
+            // picked up anyway since we can't fit sorobanTx (gaps are allowed)
+            resources.instructions = 1;
+            resources.readBytes = 1;
+            resources.writeBytes = 1;
+            resources.extendedMetaDataSizeBytes = 1;
+
+            auto smallSorobanLowFee = createSimpleDeployContractTx(
+                *app, acc4, baseFee / 10, /* refundableFee */ 1200, resources);
+
+            TxSetFrame::TxPhases invalidPhases;
+            invalidPhases.resize(2);
+            TxSetFrameConstPtr txSet = TxSetFrame::makeFromTransactions(
+                TxSetFrame::TxPhases{
+                    {tx}, {sorobanTxHighFee, smallSorobanLowFee, sorobanTx}},
+                *app, 0, 0, invalidPhases);
+
+            for (auto const& phase : invalidPhases)
+            {
+                REQUIRE(phase.empty());
+            }
+            REQUIRE(txSet->sizeTx() == 3);
+            for (auto const& tx : txSet->getTxs())
+            {
+                // smallSorobanLowFee was picked over sorobanTx to fill the gap
+                REQUIRE(tx != sorobanTx);
+            }
+        }
+        SECTION("soroban tx limits")
```

Comment:
Just a reminder this still is not done - ok to do as a followup (probably do a smoke test where we create 100-1000 random tx sets that preserve the limits?).

Reply:
thanks, I'll fix the test in a subsequent PR (maybe we'll have soroban tx limit incorporated in the network config by then).